File: info/MP2I/Cours/Cours/04-recursivite/exemples.ml
(* Exemples plus avancés *)

(** Renvoie x^n (avec n >= 0) *)
let rec exp_rap x n =
  if n > 0 then
    let x_p = exp_rap x (n/2) in (* x puissance p avec p = n/2 *)
    if n mod 2 = 0 then x_p *. x_p else x *. x_p *. x_p
  else
    1.



(** Somme des entiers jusqu'à n inclus *)
let somme_entiers n =
  let rec boucle somme i =
    if i <= n then 
      boucle (somme+i) (i+1) 
    else 
      somme
  in
  boucle 0 0


(** Renvoie k parmi n *)
let rec binom n k =
  if k = 0 || k = n then 
    1 
  else
    ( binom (n-1) k ) + ( binom (n-1) (k-1) )


(* main *)

let _ =
  let x = 2.0 in
  let n = 4 in
  let _ = Printf.printf "%f^%d = %f\n" x n (exp_rap x n) in

  let n = 5 in
  let _ = Printf.printf "1+2+...+%d = %d\n" n (somme_entiers n) in

  let _ = Printf.printf " binom 3 2 = %d\n" (binom 3 2) in
  Printf.printf " binom 5 2 = %d\n" (binom 5 2)

File: info/MP2I/Cours/04-recursivite/exemples.c
#include <stdlib.h>
#include <stdbool.h>
#include <stdio.h>
#include <assert.h>


/* Introduction */

/** Renvoie x^n. */
double exp_naif(double x, unsigned n) {
  if (n > 0) {
    return x * exp_naif(x, n-1);
  } else { // n == 0
    return 1;
  }
}



/* Exemples plus avancés */

/** Renvoie x^n. */
double exp_rap(double x, unsigned n) {
  if (n > 0) {
    int x_p = exp_rap(x, n/2); // x puissance p avec p = n/2
    if (n % 2 == 0) { return x_p * x_p; } 
    else { return x * x_p * x_p; }
  } 
  else {
    return 1;
  }
}


/** Somme des entiers jusqu'à n inclus. */
int somme_entiers(int n) {
  int i = 0;
  int somme = 0;
  while (i <= n) {
    somme = somme + i;
    i = i +1;
  }
  return i;
}


int boucle(int somme, int i, int n) {
	if (i <= n) {
		return boucle(somme+i, i+1, n);
	} else {
		return somme;
	}
}

int somme_rec(int n) {
	return boucle(0, 0, n);
}




/* main */

int main(void) {
  
  double x = 2.0;
  unsigned n = 4;
  printf("%lf^%u = %lf\n", x, n, exp_naif(x, n));

  return EXIT_SUCCESS;
}
